src/powerfit_em/correlators/gpu.py

```diff
@@ -0,0 +1,319 @@
+i32 = np.int32
+
+@dataclass
+class GPUVars:
+    target: ClArray
+    template: ClArray
+    mask: ClArray
+    lcc_mask: ClArray
+    target2: ClArray
+    rot_template: ClArray
+    rot_mask: ClArray
+    rot_mask2: ClArray
+    gcc: ClArray
+    ave: ClArray
+    ave2: ClArray
+    lcc: ClArray
+    rot: ClArray
+
+
+@dataclass
+class GPUVarsFT:
+    """Fourier transformed (complex) arrays."""
+    target: ClArray
+    target2: ClArray
+    template: ClArray
+    mask: ClArray
+    mask2: ClArray
+    ave: ClArray
+    ave2: ClArray
+    lcc: ClArray
+    gcc: ClArray
+
+
+def laplace_filter(array):
+    """Laplace transform"""
+    return laplace(array, mode='wrap')
+
+
+def get_lcc_mask(target: np.ndarray) -> np.ndarray:
+    return (target > target.max() * 0.05).astype(np.uint8)
+
+
+def get_ft_shape(target: np.ndarray) -> tuple:
+    """Returns shape of fourier transformed target."""
+    return target.shape[:-1] + (target.shape[-1] // 2 + 1,)
+
+
+def init_gpu_vars(
+    queue: cl.CommandQueue, target: np.ndarray, template: np.ndarray, mask: np.ndarray, laplace: bool,
+):
+    lcc_mask = get_lcc_mask(target)
+    _t = laplace_filter(target) if laplace else target
+    zeros = np.zeros(target.shape, f32)
+    gpu_vars = GPUVars(
+        target = cl_array.to_device(queue, _t.astype(f32)),
+        template = cl.image_from_array(queue.context, template.astype(f32)),
+        mask = cl.image_from_array(queue.context, mask.astype(f32)),
+        lcc_mask = cl_array.to_device(queue, lcc_mask.astype(i32)),
+        target2 = cl_array.to_device(queue, zeros),
+        rot_template = cl_array.to_device(queue, zeros),
+        rot_mask = cl_array.to_device(queue, zeros),
+        rot_mask2 = cl_array.to_device(queue, zeros),
+        gcc = cl_array.to_device(queue, zeros),
+        ave = cl_array.to_device(queue, zeros),
+        ave2 = cl_array.to_device(queue, zeros),
+        lcc = cl_array.to_device(queue, zeros),
+        rot = cl_array.to_device(queue, np.zeros(target.shape, i32)),
+    )
+    zeros_ft = np.zeros(get_ft_shape(target), dtype=np.complex64)
+    gpu_vars_ft = GPUVarsFT(
+        target = cl_array.to_device(queue, zeros_ft),
+        target2 = cl_array.to_device(queue, zeros_ft),
+        template = cl_array.to_device(queue, zeros_ft),
+        mask = cl_array.to_device(queue, zeros_ft),
+        mask2 = cl_array.to_device(queue, zeros_ft),
+        ave = cl_array.to_device(queue, zeros_ft),
+        ave2 = cl_array.to_device(queue, zeros_ft),
+        lcc = cl_array.to_device(queue, zeros_ft),
+        gcc = cl_array.to_device(queue, zeros_ft),
+    )
+    return gpu_vars, gpu_vars_ft
+
+
+def generate_kernels(queue: cl.CommandQueue, target: np.ndarray):
+    kernel_values = {
+        'shape_x': target.shape[2],
+        'shape_y': target.shape[1],
+        'shape_z': target.shape[0],
+        'llength': i32(min(target.shape) // 2),
+    }
+    return CLKernels(queue.context, kernel_values)
+
+
+class GPUCorrelator:
```

Comment:
@sverhoeven any comments on this new structure before I continue? Note that I still have to finish up the docstrings.

The tests fail because I broke how the powerfitter class works (will fix later), running powerfit on gpu works though.

Reply:
I have added this API in as a `set_template` method. This allows for trying to fit a different structure on the same already initialized target.

Reply:
We can use this in combination with some "queue pool" to fit a lot of different templates in parallel to a single target.